File: src/main/java/com/oycm/algorithm/f/prefix_sum_hash/Solution_16.scala
package com.oycm.algorithm.f.prefix_sum_hash

object Solution_16 {

  /**
   * 1590. 使数组和能被 P 整除 2039
   * https://leetcode.cn/problems/make-sum-divisible-by-p/description/
   *
   * 移除最短子数组，使得 剩余元素和能被 p 整除，不允许将整个子数组移除。如没有则 返回 -1
   *
   * @param nums nums(i) [1, 10 pow 9], n in [1, 10 pow 5]
   * @param p    [1, 10 pow 9]
   * @return
   */
  def minSubarray(nums: Array[Int], p: Int): Int = {
    /*
    nums 数组和 记为 sums(n), 需要移除子数组记为 [l, r), 其和可记为 sums(r) - sums(l)
    (sums(n) - (sums(r) - sums(l))) % p == 0
    sums(n) % p - sums(r) % p + sums(l) % p == 0

    sums(r) % p = sums(n) % p + sums(l) % p

    (sums(r) % p - sums(n) % p + p) % p = sums(l) % p

    先计算整个 sums % p, 遍历 sums(r) 的 过程中，查找 是否存在 sums(r) % p - sums(n) % p 的 sums(l)
    这里有减法，取模需要注意负数的情况
    */
    val n = nums.length
    var ans = n
    var sum = 0
    for (num <- nums) {
      sum = (sum + num % p) % p
    }
    // 整个数组已经符合条件
    if (sum == 0) {
      return 0
    }
    val map = scala.collection.mutable.Map[Int, Int]()
    var s = 0
    for (i <- nums.indices) {
      // 求最短，要不算 更新 相同的 sum index
      map(s) = i
      // 计算 sums(r)
      s = (s + nums(i) % p) % p
      // 更新答案 r 未 + 1
      ans = Math.min(ans, i - map.getOrElse((s - sum + p) % p, i - n) + 1)
    }
    if (ans == n) ans = -1
    ans
  }

  def main(args: Array[String]): Unit = {
    println(minSubarray(Array(3, 1, 4, 2), 6))
    println(minSubarray(Array(3, 1, 4, 2), 1))
  }
}

File: src/main/java/com/oycm/month2026/may/Solution_16.java
package com.oycm.month2026.may;

public class Solution_16 {

    /**
     * 154. <a href="https://leetcode.cn/problems/find-minimum-in-rotated-sorted-array-ii/description/">寻找旋转排序数组中的最小值 II</a>
     *
     * @param nums
     * @return
     */
    public int findMin(int[] nums) {
        /*
        已知一个长度为 n 的数组，预先按照升序排列，经由 1 到 n 次 旋转 后，得到输入数组。
        例如，原数组 nums = [0,1,2,4,5,6,7] 在变化后可能得到：
            若旋转 4 次，则可以得到 [4,5,6,7,0,1,2]
            若旋转 7 次，则可以得到 [0,1,2,4,5,6,7]
        数组 [a[0], a[1], a[2], ..., a[n-1]] 旋转一次 的结果为数组 [a[n-1], a[0], a[1], a[2], ..., a[n-2]]。
        可能存在 重复 元素值的数组 nums，它原来是一个升序排列的数组，并按上述情形进行了多次旋转。
        找出数组中的最小元素
         */
        /*
        题解思路：这里与 153 的区别是有重复元素，如果 nums[mid] == nums[n-1] 就无法判断答案在左边还是再右边。
        当 nums[mid] == nums[n-1] 就去掉 n-1 这个元素，问题变成了 n-1 的子问题。
        这里会不会去掉最小值呢？不会
            因为 nums[mid] == nums[n-1]，如果是最小值，则最小值还在数组中；
            如果不是最小值，则排除了一个错误答案
         */
        int l = -1;
        int r = nums.length - 1;
        while (l + 1 < r) {
            int mid = (l + r) >> 1;
            if (nums[mid] == nums[r]) {
                r--;
            } else if (nums[mid] > nums[r]) {
                l = mid;
            } else {
                r = mid;
            }
        }
        return nums[r];
    }

}
